dace/codegen/targets/fpga.py

```diff
@@ -0,0 +1,1417 @@
+                    "}} else {{\n{it} += {step};\n}}\n".format(
+                        it=it, step=r[2]))
+            if len(cond) > 0:
+                callsite_stream.write("}\n")
+
+        self._cpu_codegen._generate_MapExit(sdfg, dfg, state_id, node,
+                                            function_stream, callsite_stream)
+
+    def _generate_Reduce(self, sdfg, dfg, state_id, node, function_stream,
+                         callsite_stream):
+
+        end_braces = 0
+
+        axes = node.axes
+        input_memlet = dfg.in_edges(node)[0].data
+        src_data = sdfg.arrays[input_memlet.data]
+        output_edge = dfg.out_edges(node)[0]
+        output_memlet = output_edge.data
+        dst_data = sdfg.arrays[output_memlet.data]
+
+        output_type = self.make_vector_type(dst_data.dtype,
+                                            output_memlet.veclen, False)
+
+        # If axes were not defined, use all input dimensions
+        input_dims = input_memlet.subset.dims()
+        output_dims = output_memlet.subset.data_dims()
+        if axes is None:
+            axes = tuple(range(input_dims))
+        output_axes = [a for a in range(input_dims) if a not in axes]
+
+        # Obtain variable names per output and reduction axis
+        axis_vars = []
+        unroll_dim = []
+        octr = 0
+        for d in range(input_dims):
+            if d in axes:
+                axis_vars.append('__i%d' % d)
+            else:
+                axis_vars.append('__o%d' % octr)
+                octr += 1
+            if ((isinstance(src_data, dace.data.Stream)
+                 and src_data.is_stream_array()) or
+                    (isinstance(src_data, dace.data.Array) and
+                     src_data.storage == dace.dtypes.StorageType.FPGA_Registers)):
+                # Unroll reads from registers and stream arrays
+                unroll_dim.append(True)
+            else:
+                unroll_dim.append(False)
+
+        # We want to pipeline the last non-unrolled dimension
+        pipeline_dim = -1
+        for i in itertools.chain(axes, output_axes):
+            if not unroll_dim[i]:
+                pipeline_dim = i
+
+        if node.identity is not None:
+            identity = cpu.sym2cpp(node.identity)
+        else:
+            identity = None
+
+        # Determine reduction type
+        reduction_type = operations.detect_reduction_type(node.wcr)
+        if reduction_type == dace.dtypes.ReductionType.Custom:
+            raise NotImplementedError("Custom reduction for FPGA is NYI")
+
+        # Initialize accumulator variable if we're collapsing to a single value
+        all_axes_collapsed = (len(axes) == input_dims)
+        if all_axes_collapsed:
+            accumulator = "_{}_accumulator".format(output_memlet.data)
+            init_value = ""
+            if identity is not None:
+                init_value = " = " + identity
+            elif reduction_type == dace.dtypes.ReductionType.Sum:
```

Comment:
Are you referring to the `if` branch or to the `elif`?

For the latter, that is used for single-clock cycle accumulator, useful for Intel for single precision float numbers and sum only.

Reply:
It seems hacky to have this special case here. But I'm also not sure we want to add all cases (it's awkward for min/max, for example, since you need the maximum/minimum possible value for the type).

Reply:
I know, but I don't see any other place where we could put it.
I don't want to add all cases, it is needed just for sum.